# Understanding Polynomial Features

**Author: Davy WAI**

Date: Monday 6 March 2023

Status: 100% completed

When receiving dataset, we want to understand the relationshionship between the features and the target (for a Supervised Model). But for both Supervised and Unsupervised Learning, it is fascinating to understand the geometric structure of your features. You may also need to generate polymomial features to seperate your observations into a higher dimensional space.

* The Kernel Trick with a Polynomial Kernel helps you do this:
  * 📺 [Support Vector Machines Part 2: The Polynomial Kernel (Part 2 of 3) - StatQuest with Josh Starmer](https://www.youtube.com/watch?v=Toet3EiSFcM)

* How many features are you generating ? 
  * DavyDW drew a table for $p = 3$ features and $d$ = up to $2$ degrees.

<img src="https://github.com/davydw/public-pictures/blob/main/data-illustrations/generating_polynomial_features.png?raw=true" width=600 height=400>

But can we generalise the number of generated features for a higher number of features and a higher degree ? Probably. Let's go through it together.

In [1]:
import numpy as np
import pandas as pd
pd.options.display.max_columns=None

from sklearn.preprocessing import PolynomialFeatures

In [2]:
n = 1 # the number of rows does not matter, we wanna play with features and degrees

## 🥋 (1) Playing with degrees

### 🐣 Degree 1

There is nothing to generate: you have $p$ features: 
- $X_1, X_2, ..., X_p$

In [3]:
p = 4 # features
d = 1 # degree

print(f"Playing with {p} feature(s)...")
print(f"...and d={d} degree(s)")
print("-"*25)

# CREATING THE DATAFRAME WITH N = 3 ROWS AND P FEATURES
X = pd.DataFrame(np.random.rand(n,p), columns = [f"X_{k+1}" for k in range(p)])
display(X)
print(f"🐣 We went from {p} feature(s)...")

# INSTANTIATING THE POLYNOMIAL FEATURES
poly = PolynomialFeatures(degree = d)
X_poly = pd.DataFrame(poly.fit_transform(X), columns = ["intercept"]+list(poly.get_feature_names_out()[1:]))
display(X_poly)
print(f"😈 ...to {X_poly.shape[1]-1} generated feature(s) + the intercept!")

Playing with 4 feature(s)...
...and d=1 degree(s)
-------------------------


,X_1,X_2,X_3,X_4
0,0.953556,0.85007,0.378221,0.408352


🐣 We went from 4 feature(s)...


,intercept,X_1,X_2,X_3,X_4
0,1.0,0.953556,0.85007,0.378221,0.408352


😈 ...to 4 generated feature(s) + the intercept!


### 🐥 Degree 2

* Squared features: 
  * $X_1^2, X_2^2, ... X_p^2 \rightarrow p $ squared features
* Products of features:
  * $(X_{\color{blue}{1}} \cdot X_2), (X_{\color{blue}{1}} \cdot X_3), ..., (X_{\color{blue}{1}} \cdot X_p) \rightarrow {\color{blue}{p-1}}$ generated features
  * $(X_{\color{purple}{2}} \cdot X_3), (X_{\color{purple}{2}} \cdot X_4), ..., (X_{\color{purple}{2}} \cdot X_p) \rightarrow {\color{purple}{p-2}}$ generated features
  * $(X_{\color{orange}{3}} \cdot X_4), (X_{\color{orange}{3}} \cdot X_5), ..., (X_{\color{orange}{3}} \cdot X_p) \rightarrow {\color{orange}{p-3}}$ generated features  
  * ...
  * $(X_{\color{red}{p-1}} \cdot X_p) \rightarrow {\color{red}{1}}$ generated features    

In total, `d=2` generated $p$ squared features and ${\color{red}{1}} + 2 + ... + {\color{orange}{p-3}} +{\color{purple}{p-2}} + {\color{blue}{p-1}} = \frac{(p-1)p}{2}$ products of features which combined exponents sum up to `d=2`

📚 Wanna know how to sum consecutive integers? [Brilliant.org - Wiki - Sum of n, n2 or n3](https://brilliant.org/wiki/sum-of-n-n2-or-n3/)

In [4]:
p = 4 # features
d =2 # degree

print(f"Playing with {p} feature(s)...")
print(f"...and d={d} degree(s)")
print("-"*25)

# CREATING THE DATAFRAME WITH N = 3 ROWS AND P FEATURES
X = pd.DataFrame(np.random.rand(n,p), columns = [f"X{k+1}" for k in range(p)])
display(X)
print(f"🐣 We went from {p} feature(s)...")

# INSTANTIATING THE POLYNOMIAL FEATURES
poly = PolynomialFeatures(degree = d)
X_poly = pd.DataFrame(poly.fit_transform(X), columns = ["intercept"]+list(poly.get_feature_names_out()[1:]))
display(X_poly)
print(f"😈 ...to {X_poly.shape[1]-1} generated feature(s) + the intercept!")

Playing with 4 feature(s)...
...and d=2 degree(s)
-------------------------


,X1,X2,X3,X4
0,0.185213,0.997584,0.67487,0.507766


🐣 We went from 4 feature(s)...


,intercept,X1,X2,X3,X4,X1^2,X1 X2,X1 X3,X1 X4,X2^2,X2 X3,X2 X4,X3^2,X3 X4,X4^2
0,1.0,0.185213,0.997584,0.67487,0.507766,0.034304,0.184765,0.124994,0.094045,0.995175,0.673239,0.506539,0.455449,0.342676,0.257826


😈 ...to 14 generated feature(s) + the intercept!


### 🚀 Degree 3

* Cubes features: 
  * $X_1^3, X_2^3, ... X_p^3 \rightarrow p$ cubed features


* Products of features:
  * Multipling a squared feature by another one (because $2+1 =3$)
    * $(X_{\color{blue}{1}}^2 \cdot X_2), (X_{\color{blue}{1}}^2 \cdot X_3), ..., (X_{\color{blue}{1}}^2 \cdot X_p) \rightarrow {\color{blue}{p-1}}$ generated features
    * $(X_{\color{purple}{2}}^2 \cdot X_1), (X_{\color{purple}{2}}^2 \cdot X_3), ..., (X_{\color{purple}{2}}^2 \cdot X_p) \rightarrow {\color{purple}{p-1}}$ generated features
    * $(X_{\color{orange}{3}}^2 \cdot X_1), (X_{\color{orange}{3}}^2 \cdot X_2), ..., (X_{\color{orange}{3}}^2 \cdot X_p) \rightarrow {\color{orange}{p-1}}$ generated features  
    * ...
    * $(X_{\color{red}{p-1}}^2 \cdot X_1), (X_{\color{red}{p-1}}^2 \cdot X_2), ..., (X_{\color{red}{p-1}}^2 \cdot X_p) \rightarrow {\color{red}{p-1}}$ generated features  generated features    
    * $(X_{\color{green}{p}}^2 \cdot X_1), (X_{\color{green}{p}}^2 \cdot X_2), ..., (X_{\color{green}{p}}^2 \cdot X_{p-1}) \rightarrow {\color{green}{p-1}}$ generated features  generated features        
    * $\rightarrow p(p-1)$ generated features which follow the pattern $X_i^2 \cdot X_j$
  * Choosing $3$ distinct features among $p$ because $1+1+1 = 3$
    * $\rightarrow \binom{p}{3} = \frac{p!}{3!(p-3)!}$ generatures features which follow the pattern $X_i \cdot X_j \cdot X_k$

In [5]:
p = 4 # features
d = 3 # degree

print(f"Playing with {p} feature(s)...")
print(f"...and d={d} degree(s)")
print("-"*25)

# CREATING THE DATAFRAME WITH N = 3 ROWS AND P FEATURES
X = pd.DataFrame(np.random.rand(n,p), columns = [f"X{k+1}" for k in range(p)])
display(X)
print(f"🐣 We went from {p} feature(s)...")

# INSTANTIATING THE POLYNOMIAL FEATURES
poly = PolynomialFeatures(degree = d)
X_poly = pd.DataFrame(poly.fit_transform(X), columns = ["intercept"]+list(poly.get_feature_names_out()[1:]))
display(X_poly)
print(f"😈 ...to {X_poly.shape[1]-1} generated feature(s) + the intercept!")

Playing with 4 feature(s)...
...and d=3 degree(s)
-------------------------


,X1,X2,X3,X4
0,0.337372,0.528122,0.433548,0.814389


🐣 We went from 4 feature(s)...


,intercept,X1,X2,X3,X4,X1^2,X1 X2,X1 X3,X1 X4,X2^2,X2 X3,X2 X4,X3^2,X3 X4,X4^2,X1^3,X1^2 X2,X1^2 X3,X1^2 X4,X1 X2^2,X1 X2 X3,X1 X2 X4,X1 X3^2,X1 X3 X4,X1 X4^2,X2^3,X2^2 X3,X2^2 X4,X2 X3^2,X2 X3 X4,X2 X4^2,X3^3,X3^2 X4,X3 X4^2,X4^3
0,1.0,0.337372,0.528122,0.433548,0.814389,0.11382,0.178174,0.146267,0.274752,0.278913,0.228967,0.430097,0.187964,0.353077,0.663229,0.0384,0.060111,0.049346,0.092694,0.094098,0.077247,0.145103,0.063414,0.119118,0.223755,0.1473,0.120922,0.227144,0.099268,0.186468,0.350266,0.081492,0.153076,0.287542,0.540126


😈 ...to 34 generated feature(s) + the intercept!


### 🤯 Degrees 4 and 5

#### 👩🏻‍💻 Code

Let's try $d = 5 $ degrees and $p=4$ features by coding and let's then retrieve the number of generated features.

In [6]:
p = 4 # features
d = 5 # degree

print(f"Playing with {p} feature(s)...")
print(f"...and d={d} degree(s)")
print("-"*25)

# CREATING THE DATAFRAME WITH N = 3 ROWS AND P FEATURES
X = pd.DataFrame(np.random.rand(n,p), columns = [f"X{k+1}" for k in range(p)])
display(X)
print(f"🐣 We went from {p} feature(s)...")

# INSTANTIATING THE POLYNOMIAL FEATURES
poly = PolynomialFeatures(degree = d)
X_poly = pd.DataFrame(poly.fit_transform(X), columns = ["intercept"]+list(poly.get_feature_names_out()[1:]))
display(X_poly)
print(f"😈 ...to {X_poly.shape[1]-1} generated feature(s) + the intercept!")

Playing with 4 feature(s)...
...and d=5 degree(s)
-------------------------


,X1,X2,X3,X4
0,0.528924,0.929869,0.865712,0.41613


🐣 We went from 4 feature(s)...


,intercept,X1,X2,X3,X4,X1^2,X1 X2,X1 X3,X1 X4,X2^2,X2 X3,X2 X4,X3^2,X3 X4,X4^2,X1^3,X1^2 X2,X1^2 X3,X1^2 X4,X1 X2^2,X1 X2 X3,X1 X2 X4,X1 X3^2,X1 X3 X4,X1 X4^2,X2^3,X2^2 X3,X2^2 X4,X2 X3^2,X2 X3 X4,X2 X4^2,X3^3,X3^2 X4,X3 X4^2,X4^3,X1^4,X1^3 X2,X1^3 X3,X1^3 X4,X1^2 X2^2,X1^2 X2 X3,X1^2 X2 X4,X1^2 X3^2,X1^2 X3 X4,X1^2 X4^2,X1 X2^3,X1 X2^2 X3,X1 X2^2 X4,X1 X2 X3^2,X1 X2 X3 X4,X1 X2 X4^2,X1 X3^3,X1 X3^2 X4,X1 X3 X4^2,X1 X4^3,X2^4,X2^3 X3,X2^3 X4,X2^2 X3^2,X2^2 X3 X4,X2^2 X4^2,X2 X3^3,X2 X3^2 X4,X2 X3 X4^2,X2 X4^3,X3^4,X3^3 X4,X3^2 X4^2,X3 X4^3,X4^4,X1^5,X1^4 X2,X1^4 X3,X1^4 X4,X1^3 X2^2,X1^3 X2 X3,X1^3 X2 X4,X1^3 X3^2,X1^3 X3 X4,X1^3 X4^2,X1^2 X2^3,X1^2 X2^2 X3,X1^2 X2^2 X4,X1^2 X2 X3^2,X1^2 X2 X3 X4,X1^2 X2 X4^2,X1^2 X3^3,X1^2 X3^2 X4,X1^2 X3 X4^2,X1^2 X4^3,X1 X2^4,X1 X2^3 X3,X1 X2^3 X4,X1 X2^2 X3^2,X1 X2^2 X3 X4,X1 X2^2 X4^2,X1 X2 X3^3,X1 X2 X3^2 X4,X1 X2 X3 X4^2,X1 X2 X4^3,X1 X3^4,X1 X3^3 X4,X1 X3^2 X4^2,X1 X3 X4^3,X1 X4^4,X2^5,X2^4 X3,X2^4 X4,X2^3 X3^2,X2^3 X3 X4,X2^3 X4^2,X2^2 X3^3,X2^2 X3^2 X4,X2^2 X3 X4^2,X2^2 X4^3,X2 X3^4,X2 X3^3 X4,X2 X3^2 X4^2,X2 X3 X4^3,X2 X4^4,X3^5,X3^4 X4,X3^3 X4^2,X3^2 X4^3,X3 X4^4,X4^5
0,1.0,0.528924,0.929869,0.865712,0.41613,0.279761,0.49183,0.457896,0.220101,0.864656,0.804998,0.386946,0.749456,0.360248,0.173164,0.147972,0.260141,0.242192,0.116417,0.457337,0.425783,0.204665,0.396405,0.190544,0.091591,0.804017,0.748543,0.359809,0.696896,0.334984,0.16102,0.648813,0.311871,0.14991,0.072059,0.078266,0.137595,0.128101,0.061576,0.241897,0.225207,0.108252,0.209668,0.100783,0.048444,0.425264,0.395922,0.190312,0.368605,0.177181,0.085167,0.343173,0.164956,0.079291,0.038114,0.747631,0.696047,0.334575,0.648022,0.311491,0.149727,0.603311,0.289999,0.139397,0.067005,0.561685,0.26999,0.129779,0.062382,0.029986,0.041397,0.072777,0.067756,0.032569,0.127945,0.119117,0.057257,0.110899,0.053307,0.025623,0.224932,0.209413,0.10066,0.194964,0.093715,0.045047,0.181512,0.087249,0.041939,0.020159,0.39544,0.368156,0.176965,0.342754,0.164755,0.079194,0.319106,0.153388,0.07373,0.035441,0.297089,0.142804,0.068643,0.032995,0.01586,0.695199,0.647232,0.311111,0.602576,0.289646,0.139227,0.561,0.269661,0.129621,0.062306,0.522293,0.251056,0.120677,0.058007,0.027883,0.486257,0.233734,0.112351,0.054005,0.025959,0.012478


😈 ...to 125 generated feature(s) + the intercept!


❓ How did we go from 4 features to 125 features ❓

Let's break down the `PolynomialFeatures(degrees = 4).fit_transform([X1, X2, X3, X4])` step by step.

#### 1️⃣ `X_poly_1`

In [7]:
X_poly_1 = X_poly.iloc[:,1:5]
display(X_poly_1)
display(f"# polynomial features generated with degree 1 = {X_poly_1.shape[-1]}")

,X1,X2,X3,X4
0,0.528924,0.929869,0.865712,0.41613


'# polynomial features generated with degree 1 = 4'

##### 💣 **Features to the power 1: $X_k^1$**
  * $X_1, X_2, ... X_p \rightarrow p = 4 $ features to the power of 1


* Original features $X_k$
  * $X_1, X_2, ..., X_p \rightarrow p = 4$ original features

#### 2️⃣ `X_poly_2`

In [8]:
X_poly_2 = X_poly.iloc[:,5:15]
display(X_poly_2)
display(f"# polynomial features generated with degree 2 = {X_poly_2.shape[-1]}")


,X1^2,X1 X2,X1 X3,X1 X4,X2^2,X2 X3,X2 X4,X3^2,X3 X4,X4^2
0,0.279761,0.49183,0.457896,0.220101,0.864656,0.804998,0.386946,0.749456,0.360248,0.173164


'# polynomial features generated with degree 2 = 10'

##### 💣 **Features to the power of 2: $X_k^2$** (squared features)
  * $X_1^2, X_2^2, ... X_p^2 \rightarrow p = 4 $ features to the power of 2


In [9]:
mask_2_2_0 = [(("^2" in element) and (len(element)==(2+2))) for element in list(X_poly_2.columns)]
X_poly_2_2_0 = [feature for feature, b in zip(list(X_poly_2.columns), mask_2_2_0) if b]
display(X_poly_2_2_0)
print(f"len(X_poly_2_2_0) = {len(X_poly_2_2_0)}")
print(f"len(set(X_poly_2_2_0)) = {len(set(X_poly_2_2_0))}")

['X1^2', 'X2^2', 'X3^2', 'X4^2']

len(X_poly_2_2_0) = 4
len(set(X_poly_2_2_0)) = 4


##### ✖️  Products of features $X_i \cdot X_j$

```quote
2 = 1 + 1
```

* $(X_{\color{blue}{1}} \cdot X_2), (X_{\color{blue}{1}} \cdot X_3), (X_{\color{blue}{1}} \cdot X_4) \rightarrow {\color{blue}{p-1}}$ generated features
* $(X_{\color{purple}{2}} \cdot X_3), (X_{\color{purple}{2}} \cdot X_4) \rightarrow {\color{purple}{p-2}}$ generated features
* $(X_{\color{orange}{3}} \cdot X_4) \rightarrow {\color{orange}{p-3}}$ generated features  
* ...
* $(X_{\color{red}{p-1}} \cdot X_p) \rightarrow {\color{red}{1}}$ generated features    
* ${\color{red}{1}} + 2 + ... + {\color{orange}{p-3}} +{\color{purple}{p-2}} + {\color{blue}{p-1}} = \frac{(p-1)p}{2} = \frac{(4-1) \cdot 4}{2} = \frac{(3) \cdot 4}{2} = 6$ products of features in the form $X_i \cdot X_j$



In [10]:
mask_2_1_1 = [(not("^2" in element) and (len(element)==(2+1+2))) for element in list(X_poly_2.columns)]
X_poly_2_1_1 = [feature for feature, b in zip(list(X_poly_2.columns), mask_2_1_1) if b]
display(X_poly_2_1_1)
print(f"len(X_poly_2_1_1) = {len(X_poly_2_1_1)}")
print(f"len(set(X_poly_2_1_1)) = {len(set(X_poly_2_1_1))}")

['X1 X2', 'X1 X3', 'X1 X4', 'X2 X3', 'X2 X4', 'X3 X4']

len(X_poly_2_1_1) = 6
len(set(X_poly_2_1_1)) = 6


##### 🌯 ***Wrapping up for 2*** 

In [11]:
print(f"## len(X_poly_2_2_0) = {len(X_poly_2_2_0)}")
display(X_poly_2_2_0)
print(f"## len(X_poly_2_1_1) = {len(X_poly_2_1_1)}")
display(X_poly_2_1_1)


## len(X_poly_2_2_0) = 4


['X1^2', 'X2^2', 'X3^2', 'X4^2']

## len(X_poly_2_1_1) = 6


['X1 X2', 'X1 X3', 'X1 X4', 'X2 X3', 'X2 X4', 'X3 X4']

In [12]:
total_counted_2 = len(X_poly_2_2_0) + len(X_poly_2_1_1)
total_counted_2

10

In [13]:
display(f"# polynomial features generated with degree 2 = {X_poly_2.shape[-1]}")

'# polynomial features generated with degree 2 = 10'

#### 3️⃣ `X_poly_3`

In [14]:
X_poly_3 = X_poly.iloc[:,15:35]
X_poly3_features = list(X_poly_3.columns)
display(X_poly_3)
display(f"# polynomial features generated with degree 3 = {X_poly_3.shape[-1]}")


,X1^3,X1^2 X2,X1^2 X3,X1^2 X4,X1 X2^2,X1 X2 X3,X1 X2 X4,X1 X3^2,X1 X3 X4,X1 X4^2,X2^3,X2^2 X3,X2^2 X4,X2 X3^2,X2 X3 X4,X2 X4^2,X3^3,X3^2 X4,X3 X4^2,X4^3
0,0.147972,0.260141,0.242192,0.116417,0.457337,0.425783,0.204665,0.396405,0.190544,0.091591,0.804017,0.748543,0.359809,0.696896,0.334984,0.16102,0.648813,0.311871,0.14991,0.072059


'# polynomial features generated with degree 3 = 20'

##### 💣 **Features to the power of 3: $X_k^3$** (cubed features)
  * $X_1^3, X_2^3, ... X_p^3 \rightarrow p = 4 $ features to the power of 3


In [15]:
mask_3_3_0 = [(("^3" in element) and (len(element)==(2+2))) for element in X_poly3_features]
X_poly_3_3_0 = [feature for feature, b in zip(X_poly3_features, mask_3_3_0) if b]
display(X_poly_3_3_0)
print(f"len(X_poly_3_3_0) = {len(X_poly_3_3_0)}")
print(f"len(set(X_poly_3_3_0)) = {len(set(X_poly_3_3_0))}")

['X1^3', 'X2^3', 'X3^3', 'X4^3']

len(X_poly_3_3_0) = 4
len(set(X_poly_3_3_0)) = 4


##### ✖️  **Products of features**

    * 3 = 2 + 1 (= 1+2)
    * 3 = 1 + 1 + 1

######👉 ***1 dominant feature*** $X_j^2$ and ***1 minor feature*** $X_1$

($ 3 = 2 + 1 = 1 + 2 $)

In [16]:
mask_3_2_1 = [(("^2" in element) and (len(element)==(2+2+1+2))) for element in X_poly3_features]
X_poly_3_2_1 = [feature for feature, b in zip(X_poly3_features, mask_3_2_1) if b]
display(X_poly_3_2_1)
print(f"len(X_poly_3_2_1) = {len(X_poly_3_2_1)}")
print(f"len(set(X_poly_3_2_1)) = {len(set(X_poly_3_2_1))}")

['X1^2 X2',
 'X1^2 X3',
 'X1^2 X4',
 'X1 X2^2',
 'X1 X3^2',
 'X1 X4^2',
 'X2^2 X3',
 'X2^2 X4',
 'X2 X3^2',
 'X2 X4^2',
 'X3^2 X4',
 'X3 X4^2']

len(X_poly_3_2_1) = 12
len(set(X_poly_3_2_1)) = 12


###### 👉 ***2 dominant features*** $X_j, X_k$ and ***1 minor feature*** $X_l$







In [17]:
mask_3_1_1_1 = [(not("^2" in element) and not("^3" in element) and (len(element)==(2+1+2+1+2))) for element in X_poly3_features]
X_poly_3_1_1_1 = [feature for feature, b in zip(X_poly3_features, mask_3_1_1_1) if b]
display(X_poly_3_1_1_1)
print(f"len(X_poly_3_1_1_1) = {len(X_poly_3_1_1_1)}")
print(f"len(set(X_poly_3_1_1_1)) = {len(set(X_poly_3_1_1_1))}")

['X1 X2 X3', 'X1 X2 X4', 'X1 X3 X4', 'X2 X3 X4']

len(X_poly_3_1_1_1) = 4
len(set(X_poly_3_1_1_1)) = 4


##### 🌯 ***Wrapping up for 3*** 

In [18]:
print(f"### len(X_poly_3_3_0) = {len(X_poly_3_3_0)}")
display(X_poly_3_3_0)
print(f"### len(X_poly_3_2_1) = {len(X_poly_3_2_1)}")
display(X_poly_3_2_1)
print(f"### len(X_poly_3_1_1_1) = {len(X_poly_3_1_1_1)}")
display(X_poly_3_1_1_1)

### len(X_poly_3_3_0) = 4


['X1^3', 'X2^3', 'X3^3', 'X4^3']

### len(X_poly_3_2_1) = 12


['X1^2 X2',
 'X1^2 X3',
 'X1^2 X4',
 'X1 X2^2',
 'X1 X3^2',
 'X1 X4^2',
 'X2^2 X3',
 'X2^2 X4',
 'X2 X3^2',
 'X2 X4^2',
 'X3^2 X4',
 'X3 X4^2']

### len(X_poly_3_1_1_1) = 4


['X1 X2 X3', 'X1 X2 X4', 'X1 X3 X4', 'X2 X3 X4']

In [19]:
total_count_3 = len(X_poly_3_3_0) + len(X_poly_3_2_1) + len(X_poly_3_1_1_1)
total_count_3

20

In [20]:
display(f"# polynomial features generated with degree 3 = {X_poly_3.shape[-1]}")

'# polynomial features generated with degree 3 = 20'

#### 4️⃣ `X_poly_4`

In [21]:
X_poly_4 = X_poly.iloc[:,35:70]
X_poly_4_features = list(X_poly_4.columns)
display(X_poly_4)
display(f"# polynomial features generated with degree 4 = {X_poly_4.shape[-1]}")


,X1^4,X1^3 X2,X1^3 X3,X1^3 X4,X1^2 X2^2,X1^2 X2 X3,X1^2 X2 X4,X1^2 X3^2,X1^2 X3 X4,X1^2 X4^2,X1 X2^3,X1 X2^2 X3,X1 X2^2 X4,X1 X2 X3^2,X1 X2 X3 X4,X1 X2 X4^2,X1 X3^3,X1 X3^2 X4,X1 X3 X4^2,X1 X4^3,X2^4,X2^3 X3,X2^3 X4,X2^2 X3^2,X2^2 X3 X4,X2^2 X4^2,X2 X3^3,X2 X3^2 X4,X2 X3 X4^2,X2 X4^3,X3^4,X3^3 X4,X3^2 X4^2,X3 X4^3,X4^4
0,0.078266,0.137595,0.128101,0.061576,0.241897,0.225207,0.108252,0.209668,0.100783,0.048444,0.425264,0.395922,0.190312,0.368605,0.177181,0.085167,0.343173,0.164956,0.079291,0.038114,0.747631,0.696047,0.334575,0.648022,0.311491,0.149727,0.603311,0.289999,0.139397,0.067005,0.561685,0.26999,0.129779,0.062382,0.029986


'# polynomial features generated with degree 4 = 35'

##### 💣 **Features to the power of 4: $X_k^4$**
$X_1^4, X_2^4, ... X_p^4 \rightarrow p = 4 $ features to the power of 4


In [22]:
mask_4_4_0 = ["^4" in element for element in X_poly_4_features]
X_poly_4_4_0 = [feature for feature, b in zip(X_poly_4_features, mask_4_4_0) if b]
display(X_poly_4_4_0)
print(f"len(X_poly_4_4_0) = {len(X_poly_4_4_0)}")
print(f"len(set(X_poly_4_4_0)) = {len(set(X_poly_4_4_0))}")

['X1^4', 'X2^4', 'X3^4', 'X4^4']

len(X_poly_4_4_0) = 4
len(set(X_poly_4_4_0)) = 4


##### ✖️  **Products of features**

    * 4 = 3 + 1 (= 1 + 3)
    * 4 = 2 + 2 
    * 4 = 2 + 1 + 1 (= 1 + 2 + 1 = 1 + 1 + 2)

######👉 ***1 dominant feature*** $X_j^3$ and ***1 minor feature*** $X_k$

$4 = 3 + 1 = 1 + 3 $

In [23]:
mask_4_3_1 = ["^3" in element for element in X_poly_4_features]
X_poly_4_3_1 = [feature for feature, b in zip(X_poly_4_features, mask_4_3_1) if b]
display(X_poly_4_3_1)
print(f"len(X_poly_4_3_1) = {len(X_poly_4_3_1)}")
print(f"len(set(X_poly_4_3_1)) = {len(set(X_poly_4_3_1))}")

['X1^3 X2',
 'X1^3 X3',
 'X1^3 X4',
 'X1 X2^3',
 'X1 X3^3',
 'X1 X4^3',
 'X2^3 X3',
 'X2^3 X4',
 'X2 X3^3',
 'X2 X4^3',
 'X3^3 X4',
 'X3 X4^3']

len(X_poly_4_3_1) = 12
len(set(X_poly_4_3_1)) = 12


###### 👉 ***2 dominant features*** $X_j^2, X_k^2$ 

In [24]:
mask_4_2_2 = [(("^2" in element) and (len(element)==(2+2+1+2+2))) for element in X_poly_4_features]
X_poly_4_2_2 = [feature for feature, b in zip(X_poly_4_features, mask_4_2_2) if b]
display(X_poly_4_2_2)
print(f"len(X_poly_4_2_2) = {len(X_poly_4_2_2)}")
print(f"len(set(X_poly_4_2_2)) = {len(set(X_poly_4_2_2))}")

['X1^2 X2^2', 'X1^2 X3^2', 'X1^2 X4^2', 'X2^2 X3^2', 'X2^2 X4^2', 'X3^2 X4^2']

len(X_poly_4_2_2) = 6
len(set(X_poly_4_2_2)) = 6


###### 👉***1 dominant feature*** $X_j^2$ and ***2 minor features*** $X_j, X_k$







($ 4 = 2 + 1 + 1 = 1 + 2 + 1 = 1 + 1 + 2 $)

In [25]:
mask_4_2_1_1 = [((("^2") in element)  and (not ("^3") in element) and (len(element)== (2+2+1+2+1+2))) for element in X_poly_4_features]
X_poly_4_2_1_1 = [feature for feature, b in zip(X_poly_4_features, mask_4_2_1_1) if b]
display(X_poly_4_2_1_1)
print(f"len(X_poly_4_2_1_1) = {len(X_poly_4_2_1_1)}")
print(f"len(set(X_poly_4_2_1_1)) = {len(set(X_poly_4_2_1_1))}")

['X1^2 X2 X3',
 'X1^2 X2 X4',
 'X1^2 X3 X4',
 'X1 X2^2 X3',
 'X1 X2^2 X4',
 'X1 X2 X3^2',
 'X1 X2 X4^2',
 'X1 X3^2 X4',
 'X1 X3 X4^2',
 'X2^2 X3 X4',
 'X2 X3^2 X4',
 'X2 X3 X4^2']

len(X_poly_4_2_1_1) = 12
len(set(X_poly_4_2_1_1)) = 12


###### 👉 ***4 dominant features*** $X_j, X_k, X_l, X_m$ and ***no minor feature***.







In [26]:
mask_4_1_1_1_1 = [(not("^4" in element) and not("^3" in element) and not("^2" in element) and (len(element)==(2+1+2+1+2+1+2))) for element in X_poly_4_features]
X_poly_4_1_1_1_1 = [feature for feature, b in zip(X_poly_4_features, mask_4_1_1_1_1) if b]
display(X_poly_4_1_1_1_1)
print(f"len(X_poly_4_1_1_1_1) = {len(X_poly_4_1_1_1_1)}")
print(f"len(set(X_poly_4_1_1_1_1)) = {len(set(X_poly_4_1_1_1_1))}")

['X1 X2 X3 X4']

len(X_poly_4_1_1_1_1) = 1
len(set(X_poly_4_1_1_1_1)) = 1


##### 🌯 ***Wrapping up for 4*** 

In [27]:
print(f"#### len(X_poly_4_4_0) = {len(X_poly_4_4_0)}")
display(X_poly_4_4_0)
print(f"#### len(X_poly_4_3_1) = {len(X_poly_4_3_1)}")
display(X_poly_4_3_1)
print(f"#### len(X_poly_4_2_2) = {len(X_poly_4_2_2)}")
display(X_poly_4_2_2)
print(f"#### len(X_poly_4_2_1_1) = {len(X_poly_4_2_1_1)}")
display(X_poly_4_2_1_1)
print(f"#### len(X_poly_4_1_1_1_1) = {len(X_poly_4_1_1_1_1)}")
display(X_poly_4_1_1_1_1)

#### len(X_poly_4_4_0) = 4


['X1^4', 'X2^4', 'X3^4', 'X4^4']

#### len(X_poly_4_3_1) = 12


['X1^3 X2',
 'X1^3 X3',
 'X1^3 X4',
 'X1 X2^3',
 'X1 X3^3',
 'X1 X4^3',
 'X2^3 X3',
 'X2^3 X4',
 'X2 X3^3',
 'X2 X4^3',
 'X3^3 X4',
 'X3 X4^3']

#### len(X_poly_4_2_2) = 6


['X1^2 X2^2', 'X1^2 X3^2', 'X1^2 X4^2', 'X2^2 X3^2', 'X2^2 X4^2', 'X3^2 X4^2']

#### len(X_poly_4_2_1_1) = 12


['X1^2 X2 X3',
 'X1^2 X2 X4',
 'X1^2 X3 X4',
 'X1 X2^2 X3',
 'X1 X2^2 X4',
 'X1 X2 X3^2',
 'X1 X2 X4^2',
 'X1 X3^2 X4',
 'X1 X3 X4^2',
 'X2^2 X3 X4',
 'X2 X3^2 X4',
 'X2 X3 X4^2']

#### len(X_poly_4_1_1_1_1) = 1


['X1 X2 X3 X4']

In [28]:
total_counted_4 = len(X_poly_4_4_0) + len(X_poly_4_3_1) + len(X_poly_4_2_2) + len(X_poly_4_2_1_1) + len(X_poly_4_1_1_1_1)
total_counted_4

35

In [29]:
display(f"# polynomial features generated with degree 4 = {X_poly_4.shape[-1]}")

'# polynomial features generated with degree 4 = 35'

#### 5️⃣ `X_poly_5`

In [30]:
X_poly_5 = X_poly.iloc[:,70:]
X_poly_5_features = list(X_poly_5.columns)
display(X_poly_5)
display(f"# polynomial features generated with degree 5 = {X_poly_5.shape[-1]}")


,X1^5,X1^4 X2,X1^4 X3,X1^4 X4,X1^3 X2^2,X1^3 X2 X3,X1^3 X2 X4,X1^3 X3^2,X1^3 X3 X4,X1^3 X4^2,X1^2 X2^3,X1^2 X2^2 X3,X1^2 X2^2 X4,X1^2 X2 X3^2,X1^2 X2 X3 X4,X1^2 X2 X4^2,X1^2 X3^3,X1^2 X3^2 X4,X1^2 X3 X4^2,X1^2 X4^3,X1 X2^4,X1 X2^3 X3,X1 X2^3 X4,X1 X2^2 X3^2,X1 X2^2 X3 X4,X1 X2^2 X4^2,X1 X2 X3^3,X1 X2 X3^2 X4,X1 X2 X3 X4^2,X1 X2 X4^3,X1 X3^4,X1 X3^3 X4,X1 X3^2 X4^2,X1 X3 X4^3,X1 X4^4,X2^5,X2^4 X3,X2^4 X4,X2^3 X3^2,X2^3 X3 X4,X2^3 X4^2,X2^2 X3^3,X2^2 X3^2 X4,X2^2 X3 X4^2,X2^2 X4^3,X2 X3^4,X2 X3^3 X4,X2 X3^2 X4^2,X2 X3 X4^3,X2 X4^4,X3^5,X3^4 X4,X3^3 X4^2,X3^2 X4^3,X3 X4^4,X4^5
0,0.041397,0.072777,0.067756,0.032569,0.127945,0.119117,0.057257,0.110899,0.053307,0.025623,0.224932,0.209413,0.10066,0.194964,0.093715,0.045047,0.181512,0.087249,0.041939,0.020159,0.39544,0.368156,0.176965,0.342754,0.164755,0.079194,0.319106,0.153388,0.07373,0.035441,0.297089,0.142804,0.068643,0.032995,0.01586,0.695199,0.647232,0.311111,0.602576,0.289646,0.139227,0.561,0.269661,0.129621,0.062306,0.522293,0.251056,0.120677,0.058007,0.027883,0.486257,0.233734,0.112351,0.054005,0.025959,0.012478


'# polynomial features generated with degree 5 = 56'

##### 💣 **Features to the power 5: $X_k^5$**
  * $X_1^5, X_2^5, ... X_p^5 \rightarrow p = 4 $ features to the power of 5


In [31]:
mask_5_5_0 = ["^5" in element for element in X_poly_5_features]
X_poly_5_5_0 = [feature for feature, b in zip(X_poly_5_features, mask_5_5_0) if b]
display(X_poly_5_5_0)
print(f"len(X_poly_5_5) = {len(X_poly_5_5_0)}")
print(f"len(set(X_poly_5_5)) = {len(set(X_poly_5_5_0))}")

['X1^5', 'X2^5', 'X3^5', 'X4^5']

len(X_poly_5_5) = 4
len(set(X_poly_5_5)) = 4


##### ✖️  **Products of features**

    * 5 = 4 + 1 (= 1 + 4)
    * 5 = 3 + 2 (= 2 + 3)
    * 5 = 3 + 1 + 1 (= 1 + 3 + 1 = 1 + 1 +3)
    * 5 = 2 + 2 + 1 (= 2 + 1 + 2 = 1 + 2 + 2)

######👉 ***1 dominant feature*** $X_j^4$ and ***1 minor feature*** $X_k$

($ 5 = 4 + 1 = 1 + 4 $)

In [32]:
mask_5_4_1 = ["^4" in element for element in X_poly_5_features]
X_poly_5_4_1 = [feature for feature, b in zip(X_poly_5_features, mask_5_4_1) if b]
display(X_poly_5_4_1)
print(f"len(X_poly_5_4_1) = {len(X_poly_5_4_1)}")
print(f"len(set(X_poly_5_4_1)) = {len(set(X_poly_5_4_1))}")

['X1^4 X2',
 'X1^4 X3',
 'X1^4 X4',
 'X1 X2^4',
 'X1 X3^4',
 'X1 X4^4',
 'X2^4 X3',
 'X2^4 X4',
 'X2 X3^4',
 'X2 X4^4',
 'X3^4 X4',
 'X3 X4^4']

len(X_poly_5_4_1) = 12
len(set(X_poly_5_4_1)) = 12


###### 👉 ***1 dominant feature*** $X_j^3$ and ***1 minor feature*** $X_k^2$ 

($ 5 = 3 + 2 = 2 + 3 $)

In [33]:
mask_5_3_2 = [((("^3" and "^2" ) in element) and len(element) == (2+2+1+2+2)) for element in X_poly_5_features]
X_poly_5_3_2 = [feature for feature, b in zip(X_poly_5_features, mask_5_3_2) if b]
display(X_poly_5_3_2)
print(f"len(X_poly_5_3_2) = {len(X_poly_5_3_2)}")
print(f"len(set(X_poly_5_3_2)) = {len(set(X_poly_5_3_2))}")

['X1^3 X2^2',
 'X1^3 X3^2',
 'X1^3 X4^2',
 'X1^2 X2^3',
 'X1^2 X3^3',
 'X1^2 X4^3',
 'X2^3 X3^2',
 'X2^3 X4^2',
 'X2^2 X3^3',
 'X2^2 X4^3',
 'X3^3 X4^2',
 'X3^2 X4^3']

len(X_poly_5_3_2) = 12
len(set(X_poly_5_3_2)) = 12


###### 👉***1 dominant feature*** $X_j^3$ and ***2 minor features*** $X_j, X_k$







($ 5 = 3 + 1 + 1 = 1 + 3 + 1 = 1 + 1 + 3 $)

In [34]:
mask_5_3_1_1 = [((("^3") in element)  and (not ("^2") in element) and (len(element)== (2+2+1+2+1+2))) for element in X_poly_5_features]
X_poly_5_3_1_1 = [feature for feature, b in zip(X_poly_5_features, mask_5_3_1_1) if b]
display(X_poly_5_3_1_1)
print(f"len(X_poly_5_3_1_1) = {len(X_poly_5_3_1_1)}")
print(f"len(set(X_poly_5_3_1_1)) = {len(set(X_poly_5_3_1_1))}")

['X1^3 X2 X3',
 'X1^3 X2 X4',
 'X1^3 X3 X4',
 'X1 X2^3 X3',
 'X1 X2^3 X4',
 'X1 X2 X3^3',
 'X1 X2 X4^3',
 'X1 X3^3 X4',
 'X1 X3 X4^3',
 'X2^3 X3 X4',
 'X2 X3^3 X4',
 'X2 X3 X4^3']

len(X_poly_5_3_1_1) = 12
len(set(X_poly_5_3_1_1)) = 12


###### 👉 ***2 dominant features*** $X_j^2, X_k^2$ and ***1 minor feature*** $X_l$







($ 5 = 2 + 2 + 1 = 2 + 1 + 2 = 1 + 2 + 2 $)

In [35]:
mask_5_2_2_1 = [((("^2") in element)  and (not ("^3") in element) and (len(element)== (2+2+1+2+2+1+2))) for element in X_poly_5_features]
X_poly_5_2_2_1 = [feature for feature, b in zip(X_poly_5_features, mask_5_2_2_1) if b]
display(X_poly_5_2_2_1)
print(f"len(X_poly_5_2_2_1) = {len(X_poly_5_2_2_1)}")
print(f"len(set(X_poly_5_2_2_1)) = {len(set(X_poly_5_2_2_1))}")

['X1^2 X2^2 X3',
 'X1^2 X2^2 X4',
 'X1^2 X2 X3^2',
 'X1^2 X2 X4^2',
 'X1^2 X3^2 X4',
 'X1^2 X3 X4^2',
 'X1 X2^2 X3^2',
 'X1 X2^2 X4^2',
 'X1 X3^2 X4^2',
 'X2^2 X3^2 X4',
 'X2^2 X3 X4^2',
 'X2 X3^2 X4^2']

len(X_poly_5_2_2_1) = 12
len(set(X_poly_5_2_2_1)) = 12


###### 👉 ***3 dominant features*** $X_j, X_k, X_l$ and ***1 minor feature*** $X_m^2$







$ 5 = 1 + 1 + 1 + 2 = 1 + 1 + 2 + 1 = 1 + 2 + 1 + 1 = 2 + 1 + 1 + 1 $

In [36]:
mask_5_1_1_1_2 = [((("^2") in element)  and (not ("^3") in element) and (len(element)== (2+1+2+1+2+1+2+2))) for element in X_poly_5_features]
X_poly_5_1_1_1_2 = [feature for feature, b in zip(X_poly_5_features, mask_5_1_1_1_2) if b]
display(X_poly_5_1_1_1_2)
print(f"len(X_poly_5_1_1_1_2) = {len(X_poly_5_1_1_1_2)}")
print(f"len(set(X_poly_5_1_1_1_2)) = {len(set(X_poly_5_1_1_1_2))}")

['X1^2 X2 X3 X4', 'X1 X2^2 X3 X4', 'X1 X2 X3^2 X4', 'X1 X2 X3 X4^2']

len(X_poly_5_1_1_1_2) = 4
len(set(X_poly_5_1_1_1_2)) = 4


##### 🌯 ***Wrapping up for 5***

In [37]:
print(f"len(X_poly_5_5_0) = {len(X_poly_5_5_0)}")
print(f"len(X_poly_5_4_1) = {len(X_poly_5_4_1)}")
print(f"len(X_poly_5_3_2) = {len(X_poly_5_3_2)}")
print(f"len(X_poly_5_3_1_1) = {len(X_poly_5_3_1_1)}")
print(f"len(X_poly_5_2_2_1) = {len(X_poly_5_2_2_1)}")
print(f"len(X_poly_5_1_1_1_2) = {len(X_poly_5_1_1_1_2)}")

len(X_poly_5_5_0) = 4
len(X_poly_5_4_1) = 12
len(X_poly_5_3_2) = 12
len(X_poly_5_3_1_1) = 12
len(X_poly_5_2_2_1) = 12
len(X_poly_5_1_1_1_2) = 4


In [38]:
total_counted_5 = len(X_poly_5_5_0) + len(X_poly_5_4_1) + len(X_poly_5_3_2) + len(X_poly_5_3_1_1) + len(X_poly_5_2_2_1) + len(X_poly_5_1_1_1_2)
total_counted_5

56

In [39]:
display(f"# polynomial features generated with degree 5 = {X_poly_5.shape[-1]}")

'# polynomial features generated with degree 5 = 56'

## 🥷🏻 Generalisation 

### 🏋🏻‍♂️ Motivation

😱 If we have $p$ features and $d$ degrees, it will be very daunting to count the number of elements of the polynomial function with degree $d$...

### 🕵🏻‍♂️ Mission

🤔 What shall we do if we want to use $k$ features among the $p$ features that are available to generate a polynomial function with degree $d$ ? 


> 🎯 We need to create the following function:
> $$ X_{(1)}^{d_1} \cdot X_{(2)}^{d_2} ... X_{(k-1)}^{d_{k-1}} \cdot X_{(k)}^{d_k}  $$
> such that:
> $$d_{1} + d_{2} + ... + d_{k-1} + d_{k} = d$$

### 🔭 Taking a step back...

This is exactly what we did in our previous analysis with $d=5$ !

* 💣 $5 = 5 = d_{1}$
* ✖️ $5 = 4 + 1 = d_{1} + d_{2}$ or $(1+4)$
* ✖️ $5 = 3 + 2 = d_{1} + d_{2}$ or $(2+3)$
* ✖️ $5 = 3 + 1 + 1 = d_{1} + d_{2} + d_{3}$ 
  * or $(1+3+1)$
  * or $(1+1+3)$
* ✖️ $5 = 2 + 2 + 1 = d_{1} + d_{2} + d_{3}$
  * or $(2+1+2)$
  * or $(1+2+2)$
* ✖️ $5 = 1 + 1 + 1 + 2 = d_{1} + d_{2} + d_{3} + d_{4}$
  * or $(1+1+2+1)$ 
  * or $(1+2+1+1)$ 
  * or $(2+1+1+1)$
* ✖️ $5 = 1 + 1 + 1 + 1 + 1 = d_{1} + d_{2} + d_{3} + d_{4} + d_{5}$

### 🔢 The science of counting

##### Step 1 : choosing $k$ features among $p$

\begin{equation}
  {p\choose k} =
    \begin{cases}
      \frac{p!}{k!(p-k)!} & \text{if $0 \leq k \leq p$}\\
      0 & \text{otherwise}
    \end{cases}       
\end{equation}    



##### Step 2 : choosing $k$ degrees for the $k$ chosen features

For each feature $X_1, X_2, \dots, X_{k-1}, X_k$, you need to choose its respective degree $d_1, d_2, \dots, d_{k-1}, d_k$ such that  $d_{1} + d_{2} + \dots + d_{k-1} + d_{k} = d$

$ {d\choose d_1} \times {d-d_1\choose d_2} \times {d-(d_1+d_2)\choose d_3} \times \dots \times {d-(d_1 + d_2 + \dots + d_{k-2})\choose d_{k-1}} \times {d-(d_1 + d_2 + \dots + d_{k-1})\choose d_{k}}$

$= \frac{d!}{d_1!(d-d_1)!} \times \frac{(d-d_1)!}{d_2!(d-d_1-d_2)!} \times \frac{(d-d_1-d_2)!}{d_3!(d-d_1-d_2-d_3)!} \times \dots \times \frac{(d-d_1-d_2-\dots-d_{k-2})!}{d_{k-1}!d_k!} \times \frac{d_k!}{d_k!0!} $

$=\frac{d!}{d_1! d_2! ... d_k!}$

$= {d\choose{d_1,\dots,d_k}}$

> This number is called a **`multinomial coefficient`**!
> Read this article from Wikipedia 👉 [link](https://en.wikipedia.org/wiki/Multinomial_theorem).

##### Step 3 : How many combos with $k$ features, with a total degree $d$ ?

###### Multinomial Theorem - Generalised 

$$ \sum_{d_1 + \dots + d_k = d} {d\choose{d_1,\dots,d_k}} x_{(1)}^{d_1} \cdot x_{(2)}^{d_2} \dots x_{(k)}^{d_k}= (x_{(1)}+x_{(2)}+\dots+x_{(k)})^{d}$$

###### Multinomial Theorem with $x_{(1)} = x_{(2)} = \dots = x_{(k)} = 1$

$$ \sum_{d_1 + \dots + d_k = d} {d\choose{d_1,\dots,d_k}} = k^{d}$$

###### Number of multinomial coefficients

> The number of terms in a multinomial sum, $N_{d,p}$, is equal to the number of monomials of degree $d$ on the variables $X_1, X_2, ...X_k$:

$$N_{d,k} = {d + k -1 \choose k-1}$$

> Think about the [Stars and bars (combinatorics) strategy](https://en.wikipedia.org/wiki/Stars_and_bars_(combinatorics))


### 🚔 Fact-checking

##### ⚙️ Parameters

$p=4$ features

#### 1️⃣ `X_poly_1 verification`:

$N_{1,4} = {1 + 4 -1 \choose 4-1} = {4 \choose 3} = 4 $


Indeed, for monomials of degree $d=1$, we need to create the following function:
> $$ X_{(1)}^{d_1} \cdot X_{(2)}^{d_2} ... X_{(k-1)}^{d_{k-1}} \cdot X_{(k)}^{d_k}  $$
> such that:
> $$d_{1} + d_{2} + ... + d_{k-1} + d_{k} = 1$$

The only way to do these if to consider 

#### 2️⃣ `X_poly_2 verification`:

$N_{2,4} = {2 + 4 -1 \choose 4-1} = {5 \choose 3} = 10 $


#### 3️⃣ `X_poly_3 verification`:

$N_{3,4} = {3 + 4 -1 \choose 4-1} = {6 \choose 3} = 20 $


#### 4️⃣ `X_poly_4 verification`:

$N_{4,4} = {4 + 4 -1 \choose 4-1} = {7 \choose 3} = 35 $


#### 5️⃣ `X_poly_5 verification`:

$N_{5,4} = {5 + 4 -1 \choose 4-1} = {8 \choose 3} = 56 $


In [40]:
4+10+20+35+56

125

## 🔥🔥🔥 Conclusion

With $p$ features and $d$ degrees, the shape of `X_poly = PolynomialFeatures(degree=d).fit_transform9[X1, X2, ..., XP]` will be the following:

$$\sum_{j=1}^{d} N_{j,p} = \sum_{j=1}^{p} {j + p -1 \choose p-1}$$

## 🤗 Cynthia's question for `0405-C2-LearningCurves`

In [41]:
p = 4 # features
d = 10 # degree
n= 1 #row

print(f"Playing with {p} feature(s)...")
print(f"...and d={d} degree(s)")
print("-"*25)

# CREATING THE DATAFRAME WITH N = 3 ROWS AND P FEATURES
X = pd.DataFrame(np.random.rand(n,p), columns = [f"X{k+1}" for k in range(p)])
display(X)
print(f"🐣 We went from {p} feature(s)...")

# INSTANTIATING THE POLYNOMIAL FEATURES
poly = PolynomialFeatures(degree = d)
X_poly = pd.DataFrame(poly.fit_transform(X), columns = ["intercept"]+list(poly.get_feature_names_out()[1:]))
display(X_poly)
print(f"😈 ...to {X_poly.shape[1]-1} generated feature(s) + the intercept!")

Playing with 4 feature(s)...
...and d=10 degree(s)
-------------------------


,X1,X2,X3,X4
0,0.837057,0.249161,0.850835,0.274247


🐣 We went from 4 feature(s)...


,intercept,X1,X2,X3,X4,X1^2,X1 X2,X1 X3,X1 X4,X2^2,X2 X3,X2 X4,X3^2,X3 X4,X4^2,X1^3,X1^2 X2,X1^2 X3,X1^2 X4,X1 X2^2,X1 X2 X3,X1 X2 X4,X1 X3^2,X1 X3 X4,X1 X4^2,X2^3,X2^2 X3,X2^2 X4,X2 X3^2,X2 X3 X4,X2 X4^2,X3^3,X3^2 X4,X3 X4^2,X4^3,X1^4,X1^3 X2,X1^3 X3,X1^3 X4,X1^2 X2^2,X1^2 X2 X3,X1^2 X2 X4,X1^2 X3^2,X1^2 X3 X4,X1^2 X4^2,X1 X2^3,X1 X2^2 X3,X1 X2^2 X4,X1 X2 X3^2,X1 X2 X3 X4,X1 X2 X4^2,X1 X3^3,X1 X3^2 X4,X1 X3 X4^2,X1 X4^3,X2^4,X2^3 X3,X2^3 X4,X2^2 X3^2,X2^2 X3 X4,X2^2 X4^2,X2 X3^3,X2 X3^2 X4,X2 X3 X4^2,X2 X4^3,X3^4,X3^3 X4,X3^2 X4^2,X3 X4^3,X4^4,X1^5,X1^4 X2,X1^4 X3,X1^4 X4,X1^3 X2^2,X1^3 X2 X3,X1^3 X2 X4,X1^3 X3^2,X1^3 X3 X4,X1^3 X4^2,X1^2 X2^3,X1^2 X2^2 X3,X1^2 X2^2 X4,X1^2 X2 X3^2,X1^2 X2 X3 X4,X1^2 X2 X4^2,X1^2 X3^3,X1^2 X3^2 X4,X1^2 X3 X4^2,X1^2 X4^3,X1 X2^4,X1 X2^3 X3,X1 X2^3 X4,X1 X2^2 X3^2,X1 X2^2 X3 X4,X1 X2^2 X4^2,X1 X2 X3^3,X1 X2 X3^2 X4,X1 X2 X3 X4^2,X1 X2 X4^3,X1 X3^4,X1 X3^3 X4,X1 X3^2 X4^2,X1 X3 X4^3,X1 X4^4,X2^5,X2^4 X3,X2^4 X4,X2^3 X3^2,X2^3 X3 X4,X2^3 X4^2,X2^2 X3^3,X2^2 X3^2 X4,X2^2 X3 X4^2,X2^2 X4^3,X2 X3^4,X2 X3^3 X4,X2 X3^2 X4^2,X2 X3 X4^3,X2 X4^4,X3^5,X3^4 X4,X3^3 X4^2,X3^2 X4^3,X3 X4^4,X4^5,X1^6,X1^5 X2,X1^5 X3,X1^5 X4,X1^4 X2^2,X1^4 X2 X3,X1^4 X2 X4,X1^4 X3^2,X1^4 X3 X4,X1^4 X4^2,X1^3 X2^3,X1^3 X2^2 X3,X1^3 X2^2 X4,X1^3 X2 X3^2,X1^3 X2 X3 X4,X1^3 X2 X4^2,X1^3 X3^3,X1^3 X3^2 X4,X1^3 X3 X4^2,X1^3 X4^3,X1^2 X2^4,X1^2 X2^3 X3,X1^2 X2^3 X4,X1^2 X2^2 X3^2,X1^2 X2^2 X3 X4,X1^2 X2^2 X4^2,X1^2 X2 X3^3,X1^2 X2 X3^2 X4,X1^2 X2 X3 X4^2,X1^2 X2 X4^3,X1^2 X3^4,X1^2 X3^3 X4,X1^2 X3^2 X4^2,X1^2 X3 X4^3,X1^2 X4^4,X1 X2^5,X1 X2^4 X3,X1 X2^4 X4,X1 X2^3 X3^2,X1 X2^3 X3 X4,X1 X2^3 X4^2,X1 X2^2 X3^3,X1 X2^2 X3^2 X4,X1 X2^2 X3 X4^2,X1 X2^2 X4^3,X1 X2 X3^4,X1 X2 X3^3 X4,X1 X2 X3^2 X4^2,X1 X2 X3 X4^3,X1 X2 X4^4,X1 X3^5,X1 X3^4 X4,X1 X3^3 X4^2,X1 X3^2 X4^3,X1 X3 X4^4,X1 X4^5,X2^6,X2^5 X3,X2^5 X4,X2^4 X3^2,X2^4 X3 X4,X2^4 X4^2,X2^3 X3^3,X2^3 X3^2 X4,X2^3 X3 X4^2,X2^3 X4^3,X2^2 X3^4,X2^2 X3^3 X4,X2^2 X3^2 X4^2,X2^2 X3 X4^3,X2^2 X4^4,X2 X3^5,X2 X3^4 X4,X2 X3^3 X4^2,X2 X3^2 X4^3,X2 X3 X4^4,X2 X4^5,X3^6,X3^5 X4,X3^4 X4^2,X3^3 X4^3,X3^2 X4^4,X3 X4^5,X4^6,X1^7,X1^6 X2,X1^6 X3,X1^6 X4,X1^5 X2^2,X1^5 X2 X3,X1^5 X2 X4,X1^5 X3^2,X1^5 X3 X4,X1^5 X4^2,X1^4 X2^3,X1^4 X2^2 X3,X1^4 X2^2 X4,X1^4 X2 X3^2,X1^4 X2 X3 X4,X1^4 X2 X4^2,X1^4 X3^3,X1^4 X3^2 X4,X1^4 X3 X4^2,X1^4 X4^3,X1^3 X2^4,X1^3 X2^3 X3,X1^3 X2^3 X4,X1^3 X2^2 X3^2,X1^3 X2^2 X3 X4,X1^3 X2^2 X4^2,X1^3 X2 X3^3,X1^3 X2 X3^2 X4,X1^3 X2 X3 X4^2,X1^3 X2 X4^3,X1^3 X3^4,X1^3 X3^3 X4,X1^3 X3^2 X4^2,X1^3 X3 X4^3,X1^3 X4^4,X1^2 X2^5,X1^2 X2^4 X3,X1^2 X2^4 X4,X1^2 X2^3 X3^2,X1^2 X2^3 X3 X4,X1^2 X2^3 X4^2,X1^2 X2^2 X3^3,X1^2 X2^2 X3^2 X4,X1^2 X2^2 X3 X4^2,X1^2 X2^2 X4^3,X1^2 X2 X3^4,X1^2 X2 X3^3 X4,X1^2 X2 X3^2 X4^2,X1^2 X2 X3 X4^3,X1^2 X2 X4^4,X1^2 X3^5,X1^2 X3^4 X4,X1^2 X3^3 X4^2,X1^2 X3^2 X4^3,X1^2 X3 X4^4,X1^2 X4^5,X1 X2^6,X1 X2^5 X3,X1 X2^5 X4,X1 X2^4 X3^2,X1 X2^4 X3 X4,X1 X2^4 X4^2,X1 X2^3 X3^3,X1 X2^3 X3^2 X4,X1 X2^3 X3 X4^2,X1 X2^3 X4^3,X1 X2^2 X3^4,X1 X2^2 X3^3 X4,X1 X2^2 X3^2 X4^2,X1 X2^2 X3 X4^3,X1 X2^2 X4^4,X1 X2 X3^5,X1 X2 X3^4 X4,X1 X2 X3^3 X4^2,X1 X2 X3^2 X4^3,X1 X2 X3 X4^4,X1 X2 X4^5,X1 X3^6,X1 X3^5 X4,X1 X3^4 X4^2,X1 X3^3 X4^3,X1 X3^2 X4^4,X1 X3 X4^5,X1 X4^6,X2^7,X2^6 X3,X2^6 X4,X2^5 X3^2,X2^5 X3 X4,X2^5 X4^2,X2^4 X3^3,X2^4 X3^2 X4,X2^4 X3 X4^2,X2^4 X4^3,X2^3 X3^4,X2^3 X3^3 X4,X2^3 X3^2 X4^2,X2^3 X3 X4^3,X2^3 X4^4,X2^2 X3^5,X2^2 X3^4 X4,X2^2 X3^3 X4^2,X2^2 X3^2 X4^3,X2^2 X3 X4^4,X2^2 X4^5,X2 X3^6,X2 X3^5 X4,X2 X3^4 X4^2,X2 X3^3 X4^3,X2 X3^2 X4^4,X2 X3 X4^5,X2 X4^6,X3^7,X3^6 X4,X3^5 X4^2,X3^4 X4^3,X3^3 X4^4,X3^2 X4^5,X3 X4^6,X4^7,X1^8,X1^7 X2,X1^7 X3,X1^7 X4,X1^6 X2^2,X1^6 X2 X3,X1^6 X2 X4,X1^6 X3^2,X1^6 X3 X4,X1^6 X4^2,X1^5 X2^3,X1^5 X2^2 X3,X1^5 X2^2 X4,X1^5 X2 X3^2,X1^5 X2 X3 X4,X1^5 X2 X4^2,X1^5 X3^3,X1^5 X3^2 X4,X1^5 X3 X4^2,X1^5 X4^3,X1^4 X2^4,X1^4 X2^3 X3,X1^4 X2^3 X4,X1^4 X2^2 X3^2,X1^4 X2^2 X3 X4,X1^4 X2^2 X4^2,X1^4 X2 X3^3,X1^4 X2 X3^2 X4,X1^4 X2 X3 X4^2,X1^4 X2 X4^3,X1^4 X3^4,X1^4 X3^3 X4,X1^4 X3^2 X4^2,X1^4 X3 X4^3,X1^4 X4^4,X1^3 X2^5,X1^3 X2^4 X3,X1^3 X2^4 X4,X1^3 X2^3 

😈 ...to 1000 generated feature(s) + the intercept!


$$\sum_{j=1}^{10} N_{j,4} = \sum_{j=1}^{10} {j + 4 -1 \choose 4-1}$$

In [45]:
def number_monomials(j, d):
  numerator = np.math.factorial(j+d-1)
  denominator = np.math.factorial(j)*np.math.factorial(d-1)
  return numerator / denominator

In [ ]:
p = 4 # features
d = 10 # degrees

number_of_generated_features = []

for j in range(1,10+1):
  temp = number_monomials(j,p)
  number_of_generated_features.append(int(temp))

number_of_generated_features

[4, 10, 20, 35, 56, 84, 120, 165, 220, 286]

In [51]:
np.sum(number_of_generated_features)

1000